lib/logstash/inputs/elasticsearch/paginated_search.rb

```diff
@@ -0,0 +1,237 @@
+          body = body.merge(:sort => {"_shard_doc": "asc"}) if @query&.dig("sort").nil?
+
+          body = body.merge(:search_after => search_after) unless search_after.nil?
+          body = body.merge(:slice => {:id => slice_id, :max => @slices}) unless slice_id.nil?
+          {
+            :size => @size,
+            :body => body
+          }
+        end
+
+        def next_page(pit_id: , search_after: nil, slice_id: nil)
+          options = search_options(pit_id: pit_id, search_after: search_after, slice_id: slice_id)
+          logger.trace("search options", options)
+          @client.search(options)
+        end
+
+        def process_page(output_queue)
+          r = yield
+          r['hits']['hits'].each { |hit| @plugin.push_hit(hit, output_queue) }
+
+          has_hits = r['hits']['hits'].any?
+          search_after = r['hits']['hits'][-1]['sort'] rescue nil
+          logger.warn("Query got data but the sort value is empty") if has_hits && search_after.nil?
+          [ has_hits, search_after ]
+        end
+
+        def with_pit
+          begin
+            pit_id = retryable(PIT_JOB) { create_pit }
+            yield pit_id if pit?(pit_id)
+          ensure
+            clear(pit_id)
+          end
+        end
+
+        def search(output_queue:, slice_id: nil, pit_id:)
+          log_details = {}
+          log_details = log_details.merge({ slice_id: slice_id, slices: @slices }) unless slice_id.nil?
+          logger.info("Query start", log_details)
+
+          has_hits = true
+          search_after = nil
+
+          while has_hits && !@plugin.stop?
+            logger.debug("Query progress", log_details)
+            has_hits, search_after = process_page(output_queue) do
+              next_page(pit_id: pit_id, search_after: search_after, slice_id: slice_id)
+            end
+          end
+
+          logger.info("Query completed", log_details)
+        end
+
+        def retryable_search(output_queue)
+          with_pit do |pit_id|
+            retryable(SEARCH_AFTER_JOB) do
+              search(output_queue: output_queue, pit_id: pit_id)
+            end
+          end
+        end
+
+        def retryable_slice_search(output_queue)
+          with_pit do |pit_id|
+            @slices.times.map do |slice_id|
+              Thread.new do
+                LogStash::Util::set_thread_name("[#{@pipeline_id}]|input|elasticsearch|slice_#{slice_id}")
+                retryable(SEARCH_AFTER_JOB) do
+                  search(output_queue: output_queue, slice_id: slice_id, pit_id: pit_id)
+                end
+              end
+            end.map(&:join)
+          end
+
+          logger.trace("#{@slices} slices completed")
+        end
+
+        def clear(pit_id)
+          logger.info("close point in time")
+          @client.close_point_in_time(:body => {:id => pit_id} ) if pit?(pit_id)
+        rescue => e
+          logger.warn("Ignoring close_point_in_time exception", message: e.message, exception: e.class)
```

Comment:
If there's nothing the user can do here and we're just closing the PIT, we can make this a debug entry instead:

```suggestion
          logger.debug("Ignoring close_point_in_time exception", message: e.message, exception: e.class)
```